### PR TITLE
added reservation

### PR DESCRIPTION
**Pull request by LobovVitaliy**

### CODE REVIEW

File: src/apiGatewayUtilities/index.ts [hunk 1]
```diff
@@ -65,4 +65,36 @@ export function add500Handler(func: (event: apigateway.Request) => Promise<apiga
       return buildApiResponse(500, error);
     }
   };
+}
+
+export async function query(
```
Comment (kostya9) [inline src/apiGatewayUtilities/index.ts:70]: Let's move these to another module, database operations have little to do with API gateway

File: src/routes/properties/propertiesRoute.ts [hunk 2]
```diff
@@ -50,6 +60,11 @@ function toResponse(property: Property, toUrl: (key: string) => string): Propert
     landmarks: property.landmarks,
     opportunities: property.opportunities,
     price: property.price,
+    ratings: property.ratings,
```
Comment (kostya9) [inline src/routes/properties/propertiesRoute.ts:63]: why do we need individual ratings in API response?

Comment (kostya9) [reply src/routes/properties/propertiesRoute.ts:63]: if we don't, lets remove them

File: src/routes/properties/propertiesRoute.ts [hunk 3]
```diff
@@ -50,6 +60,11 @@ function toResponse(property: Property, toUrl: (key: string) => string): Propert
...
+    rating: property.ratings.length
+      ? Math.round(property.ratings.reduce((c, r) => c + r.rating, 0) / property.ratings.length)
```
Comment (kostya9) [inline src/routes/properties/propertiesRoute.ts:65]: the array is always not null

File: src/routes/properties/propertiesRoute.ts [hunk 4]
```diff
@@ -50,6 +60,11 @@ function toResponse(property: Property, toUrl: (key: string) => string): Propert
     landmarks: property.landmarks,
     opportunities: property.opportunities,
     price: property.price,
+    ratings: property.ratings,
+    rating: property.ratings.length
```
Comment (kostya9) [inline src/routes/properties/propertiesRoute.ts:64]: let's create a function which takes in a property and returns the rating - this is a business operation, so we need extra granularity

File: src/apiGatewayUtilities/index.ts (hunk 1 above)
Comment (LobovVitaliy) [reply src/apiGatewayUtilities/index.ts:70]: ok

File: src/routes/properties/propertiesRoute.ts (hunk 2 above)
Comment (LobovVitaliy) [reply src/routes/properties/propertiesRoute.ts:63]: ok

File: src/routes/properties/propertiesRoute.ts (hunk 4 above)
Comment (LobovVitaliy) [reply src/routes/properties/propertiesRoute.ts:64]: ok

File: src/routes/reservation/index.ts [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    const beginDate = new Date(event.queryStringParameters!.beginDate);
+    const endDate = new Date(event.queryStringParameters!.endDate);
+
+    if (beginDate < new Date()) {
+      return buildApiResponse(500, {
+        message: 'Reservation date unavailable',
+      });
+    }
+
+    const property = await dynamo.getItem({
```
Comment (kostya9) [inline src/routes/reservation/index.ts:103]: let's use the propertyMode.ts file, where the Property is properly statically typed

File: src/routes/reservation/index.ts [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+      },
+    }).promise();
+
+    if (!property.Item) {
+      return buildApiResponse(404, {
+        message: 'Property not found',
+      });
+    }
+
+    const totalRoomsNumber = Number(property.Item!.totalRoomsNumber.N) || 1;
```
Comment (kostya9) [inline src/routes/reservation/index.ts:164]: If we use the propertyModel.ts, we will receive a proper totalRoomsNumber, we will not need this fallback to 1 here

File: src/routes/properties/propertiesRoute.ts [hunk 7]
```diff
@@ -180,9 +202,17 @@ export function propertyGetById() {
...
     const property = await dbModel.findById(id);
 
-    return property
-      ? buildApiResponse(200, toResponse(property, (key) => imageUrlFormatter(key, staticDomain)))
-      : buildNotFound();
+    if (!property) {
+      return buildNotFound();
+    }
+
+    if (!property.isConfirmed) {
```
Comment (kostya9) [inline src/routes/properties/propertiesRoute.ts:209]: Let's add a check: if the user is the creator of the property - return it to the user. The creator should be able to see his non-cinfirmed properties

File: src/routes/properties/propertiesRoute.ts [hunk 8]
```diff
@@ -326,10 +356,97 @@ export function propertiesGet() {
...
+    const body = parseBody(event);
+
+    const search = await dbModel.findById(propertyId);
+
+    if (!search) {
+      return buildNotFound();
+    }
+
+    if (!search.isConfirmed) {
+      return buildApiResponse(403, {
```
Comment (kostya9) [inline src/routes/properties/propertiesRoute.ts:394]: The same as higher about creator of the property

File: src/routes/properties/propertiesRoute.ts [hunk 9]
```diff
@@ -326,10 +356,97 @@ export function propertiesGet() {
...
+    const propertyId = event.pathParameters!.id;
+
+    const search = await dbModel.findById(propertyId);
+
+    if (!search) {
+      return buildNotFound();
+    }
+
+    if (search.isConfirmed) {
+      return buildApiResponse(403, {
```
Comment (kostya9) [inline src/routes/properties/propertiesRoute.ts:436]: The same as higher about creator of the property

File: src/routes/reservation/index.ts [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+    currDate.setDate(currDate.getDate() + 1);
+  };
+
+  dates.push(new Date(endDate));
+
+  return dates;
+}
+
+function createGetLockedRoomsCountHandler(beginDate: Date, endDate: Date) {
+  return (items: DynamoDB.Types.ItemList) => {
```
Comment (kostya9) [inline src/routes/reservation/index.ts:63]: if we change this to expect an object with startDate and endDate, we can unit-test this funciton properly with different combinations of dates

File: src/routes/reservation/index.ts [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+      },
+    }).promise();
+
+    if (!reservation.Item) {
+      return buildApiResponse(404, {
+        message: 'Reservation not found',
+      });
+    }
+
+    if (reservation.Item.customerId.S !== customerId) {
```
Comment (kostya9) [inline src/routes/reservation/index.ts:228]: Nice!

File: src/routes/reservation/index.ts [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+      return buildApiResponse(400, {
+        message: 'You cannot delete someone else’s reservation',
+      });
+    }
+
+    if (new Date(String(reservation.Item.beginDate.S)) <= new Date()) {
+      return buildApiResponse(500, {
+        message: 'Reservation date has ended',
```
Comment (kostya9) [inline src/routes/reservation/index.ts:236]: Maybe you mean "the reservation duration has started" or smthing?

File: src/routes/reservation/index.ts [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+
+export function getAvailableCountReservations() {
+  const dynamo = createDynamo();
+
+  const handler = async (event: awsx.apigateway.Request) => {
+    const propertyId = event.pathParameters!.id;
+    const beginDate = new Date(event.queryStringParameters!.beginDate);
+    const endDate = new Date(event.queryStringParameters!.endDate);
+
+    if (beginDate < new Date()) {
```
Comment (kostya9) [inline src/routes/reservation/index.ts:97]: let's also check that endDate > beginDate

File: src/routes/reservation/index.ts [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+      return buildApiResponse(500, {
```
Comment (kostya9) [inline src/routes/reservation/index.ts:98]: 400

File: src/routes/reservation/index.ts [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+      },
+      Limit: 1000,
+    };
+
+    const availableRoomsCount = await query(dynamo, params)
+      .then(createGetLockedRoomsCountHandler(beginDate, endDate))
+      .then((lockedRoomsCount) => totalRoomsNumber - lockedRoomsCount);
+
+    if (availableRoomsCount < bookedRoomsNumber) {
+      return buildApiResponse(500, {
```
Comment (kostya9) [inline src/routes/reservation/index.ts:181]: 400

File: src/routes/reservation/index.ts [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+        message: `available rooms count: ${availableRoomsCount}`
```
Comment (kostya9) [inline src/routes/reservation/index.ts:182]: Please specify that there are not enough rooms

**Pull request merged**